source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+It should be also tested that VMs with TPM data created on hosts running older systems are unchanged when the VM is started on a host running the newest system.  This must be tested in the guest OS.
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+### Data persistence
+
+A very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
+
+There are several options how to provide TPM data on the VM start:
+
+- As a parameter to a VM.create API call.
+- In metadata section of the domain XML.
+- On a shared storage.
+
+The hard part is how to store updated TPM data from the host reliably, making sure that:
+
+- Data updates are not lost.
+- The updated data is consistent, i.e. it can be read by swtpm correctly on the next VM start.
+
+We cannot handle all erroneous situations (e.g. faulty hardware corrupting data) but a reasonable level of resistance against common failures should be granted.  Ideally, the following situations should be handled:
+
+- When a VM stops for any reason on a connected host, we store the latest state of the data.
+- When a host or a storage get disconnected, we store the latest known state of the data.
+- When a VM is restarted without Engine assistance, it uses the final data from the last run.
+- We don't store inconsistent data.  swtpm apparently [doesn't store changes atomically](https://github.com/stefanberger/swtpm/blob/609dfd873a31fc0402752c50d1d5cfa58949507f/src/swtpm/swtpm_nvfile.c#L492), so whether we copy swtpm data or let swtpm write to shared storage, we are at risk of race with swtpm writes on copying or when the storage gets suddenly disconnected.
+- Temporary network disconnections, lost events sent from Vdsm to Engine or temporary failures when delivering or processing Vdsm responses to Engine don't cause permanent loss of data updates.
+
+The options to deliver updated TPM data are similar to obtaining TPM data on VM start, with the important distinction that this time they are write instead of read operations, with the risk of data loss.  We can:
+
+- Send the updated data to Engine in events, current API call responses (stats, destroy) or using a new Vdsm API call for the given purpose.
+- Store the data in metadata, to be read and processed in Engine monitoring.
+- Store the data to shared storage from Vdsm.
```

Comment:
This does not make sense, requires creating special volume per vm. If we do this then there is no need to keep tpm data elsewhere, it will simply be in shared storage. This requires support from swtpm, they should be able to use block device for the data, and use direct I/O to ensure consistency.